### ISSUE TITLE
Possible fault while tuning for unusual gemm dimensions 

### ISSUE THREAD

**Issue by newling** [CNugteren/CLBlast#161]
Hello, 

I'm wondering if xgemm_direct supports any dimensions. Runs such as 
  ./clblast_tuner_xgemm_direct -m 1024 -n 1024 -k 1024
  ./clblast_tuner_xgemm_direct -m 1024 -n 2048 -k 2048
work fine, but runs such as 
  ./clblast_tuner_xgemm_direct -m 1000 -n 1000 -k 1000
  ./clblast_tuner_xgemm_direct -m 2000 -n 2000 -k 2000
fail:
`Memory access fault by GPU node-1 on address 0x901f65000. `

Ultimately I would like to tune for small, highly-skew gemm problems.

Best,
James

**Comment by CNugteren** [comment]
I think this is the same issue as reported in #140. And I actually started looking into that recently, so hopefully I'll find a solution soon (not much spare time at the moment, but will try to prioritise this).

**Comment by CNugteren** [comment]
I've just updated [CLTune](https://github.com/CNugteren/CLTune) to automatically make the global size a multiple of the local work-group size. This might not be sufficient to fix your issue (if it is indeed related to the earlier mentioned case), but it is necessary: before this update to CLTune setting `-m` to 1000 would not work with a workgroup size of 16, since 1000 is not a multiple of 16. This should be fixed now.

Could you retry with the latest CLTune (`master` branch is OK) and let me know if the issue is still present?

**Comment by newling** [comment]
Still getting the error with m=n=k=1000

**Comment by CNugteren** [comment]
Could you give me the output of an example crashing run? Could you also let me know what OS and OpenCL platform you are using?

**Comment by newling** [comment]
This seems to be an issue with [ROCm 1.5](https://github.com/RadeonOpenCompute/ROCm), as with Catalyst driver Ubuntu 16.04 direct GEMM with m=n=k=1000 works fine. I'm using Ubuntu 16.04 with a Fiji GPU. I'll update ROCm to 1.6 and let you know if whatever the compiler/runtime issue is has been resolved.

**Comment by newling** [comment]
With ROCm 1.6 it seems worse, even 2^k dimensions are failing
`./clblast_tuner_xgemm_direct -m 128 -n 128 -k 128`

[==========] Initializing on platform 0 device 0
[==========] Device name: 'gfx803' (OpenCL 1.2 )
[----------] Testing reference XgemmDirectTN
device compiler error/warning: error: <unknown>:0:0: in function XgemmDirectNN void (i32, i32, i32, float, float, float addrspace(1)*, i32, i32, float addrspace(1)*, i32, i32, float addrspace(1)*, i32, i32, i32, i32, i32): unsupported call to function XgemmDirect

Error: Creating the executable failed: Compiling LLVM IRs to executable

[   FAILED ] Kernel XgemmDirectTN failed
[   FAILED ]   catched exception: device compiler error/warning occurred ^^

It seems like ROCm is only struggling with GEMM kernels, as `./clblast_tuner_xdot`  works fine.  Anyway, I don't think this is a CLBlast issue.

**Comment by CNugteren** [comment]
OK, thanks for the info. Still, it would be good to make it work on ROCm. I'm not sure I have access to a platform on which I can test it, because support seems quite limited. I have access to Ryzen, but ROCm is a GPU thing, right? And I do have access to an Ubuntu 16.04 laptop with an AMD GPU, but I believe the GPU might be too old to make it work (HD8xxx series). I tried once to install OpenCL on it, but without success.

By the way, I suppose you tried with the latest CLBlast (1st of July or later - after #167 was fixed)? Not related probably to the compilation error you are getting. The message `unsupported call to function XgemmDirect` doesn't tell me much unfortunately. Did you report this with AMD?

**Comment by newling** [comment]
Yes, it would be good to see it work on ROCm. As far I know, ROCm is a GPU-CPU thing. It's not clear to me which GPUs are supported, although I know some older CPUs are not supported (Ryzen is fine). 

I do have the July 1 changes. I haven't reported this with AMD. B.t.w. I have seen code other than CLBlast run fine on Catalyst but have problems on ROCm. I know that ROCm is under heavy development, maybe with the next release direct GEMM will just start working?

**Comment by CNugteren** [comment]
Perhaps this issue is related to the inline keyword in the kernels. Could you remove all of them from the kernels and try again? See also [this CLTune issue for more details](https://github.com/CNugteren/CLTune/issues/49) and a recently opened issue #171 using Mesa OpenCL. At least the (quite generic) error message is the same.

**Comment by CNugteren** [comment]
This is now merged into master. Could you try to see if that fixes your issue?

**Comment by newling** [comment]
It looks good. `./clblast_tuner_xgemm_direct  -m 271 -n 293 -k 237` is [OK] for the first 41 out of 41 tests, I haven't run further.

One thing, `./clblast_tuner_xgemm_direct  -m 130 -n 293 -k 237` gives correctness [WARNING]s. I guess such kernels (too-large-tiles for the problem?) get filtered out later though.

**Comment by CNugteren** [comment]
There could indeed be correctness warnings in the tuners; those results are indeed filtered out later. As long as there are not too many warnings, this shouldn't be an issue. The tuners try to test only legal combinations, but not always everything is caught. It could even be that too many registers are used or something.

**Comment by newling** [comment]
Good, thanks for this info.

**Comment by CNugteren** [comment]
I've also run the above two commands on my machine, and I also see no warnings for the first case (`./clblast_tuner_xgemm_direct -m 271 -n 293 -k 237`) but warnings for the second case (`./clblast_tuner_xgemm_direct -m 130 -n 293 -k 237`). I'll take a look on my system to see why those configurations wouldn't work and whether this is a bug or something less important.

**Comment by CNugteren** [comment]
That last issue turned out to be a problem with the CLTune tuner, not with CLBlast itself. I've now changed the latest `master` branch of CLTune to perform ceiled divisions instead of a normal rounding division for local and global work sizes. That fixes it on my machine. Could you test it as well? Thanks!

**Comment by newling** [comment]
No more warnings for `./clblast_tuner_xgemm_direct -m 130 -n 293 -k 237` on my end, brilliant!

**Comment by CNugteren** [comment]
Good to hear! :-)

**Comment by newling** [comment]
The above discussion was for `clblast_tuner_xgemm_direct`. I am now experiencing the same problem for `clblast_tuner_xgemm`. Specifically, non-2^k square problems fail. Possibly the same fix can be applied?

**Comment by CNugteren** [comment]
The GEMM kernel makes certain assumptions on the input. In the actual routine there are some pre-processing and post-processing kernels to satisfy these assumptions. Thus, it can well be that for certain sizes strange things will happen. One important assumption is that the m, n, k values should be multiples of the work-group-size related parameters. So try to round up to the nearest multiple of 32 or perhaps 64 and see if that works.

**Comment by newling** [comment]
That makes sense. Thanks.
